util/types/convert.go

```diff
@@ -39,6 +41,232 @@ func truncateStr(str string, flen int) string {
+		err = overflow(val, tp)
+		return
+	}
+	converted = val
+	return
+}
+
+func convertToInt(val interface{}, target *FieldType) (converted int64, err error) {
+	tp := target.Tp
+	lowerBound := signedLowerBound[tp]
+	upperBound := signedUpperBound[tp]
+	switch v := val.(type) {
+	case uint64:
+		if v > uint64(upperBound) {
+			converted = upperBound
+			err = overflow(val, tp)
+			return
+		}
+		converted = int64(v)
+		return
+	case int:
+		return convertIntToInt(int64(v), lowerBound, upperBound, tp)
+	case int64:
+		return convertIntToInt(int64(v), lowerBound, upperBound, tp)
+	case float32:
+		return convertFloatToInt(float64(v), lowerBound, upperBound, tp)
+	case float64:
+		return convertFloatToInt(float64(v), lowerBound, upperBound, tp)
+	case string:
+		fval, err := StrToFloat(v)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		return convertFloatToInt(fval, lowerBound, upperBound, tp)
+	case mysql.Decimal:
+		fval, _ := v.Float64()
+		return convertFloatToInt(fval, lowerBound, upperBound, tp)
+	}
+	return 0, typeError(val, target)
+}
+
+func convertIntToUint(val int64, upperBound uint64, tp byte) (converted uint64, err error) {
+	if val < 0 {
+		converted = 0
+		err = overflow(val, tp)
+		return
+	}
+	if uint64(val) > upperBound {
+		converted = upperBound
+		err = overflow(val, tp)
+		return
+	}
+	converted = uint64(val)
+	return
+}
+
+func convertFloatToUint(val float64, upperBound uint64, tp byte) (converted uint64, err error) {
+	val = RoundFloat(val)
+	if val < 0 {
+		converted = 0
+		err = overflow(val, tp)
+		return
+	}
+	if val > float64(upperBound) {
+		converted = upperBound
+		err = overflow(val, tp)
+		return
+	}
+	converted = uint64(val)
+	return
+}
+
+func convertToUint(val interface{}, target *FieldType) (converted uint64, err error) {
+	tp := target.Tp
+	upperBound := unsignedUpperBound[tp]
+	switch v := val.(type) {
+	case uint64:
+		if v > upperBound {
+			converted = upperBound
+			err = overflow(val, tp)
+			return
+		}
+		converted = v
+		return
+	case int:
+		return convertIntToUint(int64(v), upperBound, tp)
+	case int64:
+		return convertIntToUint(int64(v), upperBound, tp)
+	case float32:
+		return convertFloatToUint(float64(v), upperBound, tp)
+	case float64:
+		return convertFloatToUint(float64(v), upperBound, tp)
+	case string:
+		fval, err := StrToFloat(v)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		return convertFloatToUint(float64(fval), upperBound, tp)
+	case mysql.Decimal:
+		fval, _ := v.Float64()
+		return convertFloatToUint(fval, upperBound, tp)
+	}
+	return 0, typeError(val, target)
+}
+
+// typeError returns error for invalid value type.
+func typeError(v interface{}, target *FieldType) error {
+	return errors.Errorf("cannot use %v (type %T) in assignment to, or comparison with, column type %s)",
+		v, v, target.String())
+}
+
+// Cast casts val to certain types and does not return error.
+func Cast(val interface{}, target *FieldType) (v interface{}) { //NTYPE
```

Comment:
This comment annotates things which had to be updated to handle newly added QL types.